src/ethereum_optimized/state_db.py

```diff
@@ -0,0 +1,322 @@
+
+        if not state.tx_restore_points:
+            state.created_accounts.clear()
+
+    def get_storage(state: State, address: Address, key: Bytes) -> U256:
+        """
+        See `state`.
+        """
+        if (
+            address in state.dirty_storage
+            and key in state.dirty_storage[address]
+        ):
+            return state.dirty_storage[address][key]
+
+        if address in state.destroyed_accounts:
+            return U256(0)
+        else:
+            return U256(state.db.get_storage(address, key))
+
+    def get_storage_original(
+        state: State, address: Address, key: Bytes
+    ) -> U256:
+        """
+        See `state`.
+        """
+        if address in state.created_accounts:
+            return U256(0)
+        else:
+            return U256(state.db.get_storage(address, key))
+
+    def set_storage(
+        state: State, address: Address, key: Bytes, value: U256
+    ) -> None:
+        """
+        See `state`.
+        """
+        if address not in state.dirty_accounts:
+            state.dirty_accounts[address] = get_account_optional(
+                state, address
+            )
+        if address not in state.dirty_storage:
+            state.dirty_storage[address] = {}
+        if key not in state.dirty_storage[address]:
+            state.journal.append((address, key, Unmodified))
+        else:
+            state.journal.append(
+                (address, key, state.dirty_storage[address][key])
+            )
+        state.dirty_storage[address][key] = value
+
+    def get_account_optional(
+        state: State, address: Address
+    ) -> Optional[Account_]:
+        """
+        See `state`.
+        """
+        if address in state.dirty_accounts:
+            return state.dirty_accounts[address]
+        account = state.db.get_account_optional(address)
+        if account is not None:
+            return Account(Uint(account[0]), U256(account[1]), account[2])
+        else:
+            return None
+
+    def set_account(
+        state: State, address: Address, account: Optional[Account_]
+    ) -> None:
+        """
+        See `state`.
+        """
+        if address not in state.dirty_accounts:
+            state.journal.append((address, Unmodified))
+        if address in state.dirty_accounts:
+            state.journal.append((address, state.dirty_accounts[address]))
+        state.dirty_accounts[address] = account
+
+    def destroy_storage(state: State, address: Address) -> None:
+        """
+        See `state`.
+        """
+        state.journal.append((address, state.dirty_storage.pop(address, {})))
+        state.destroyed_accounts.add(address)
+        set_account(state, address, get_account_optional(state, address))
+
+    def mark_account_created(state: State, address: Address) -> None:
+        """
+        See `state`.
+        """
+        state.created_accounts.add(address)
+
+    return locals()
```

Comment:
Quickly thinking about `locals`, could you maybe use a class or module to hold all the replacement functions/variables and iterate through that?

Something like:

```python
class Patch:
    @dataclass
    class State:
        ...

    def close_state(...):
        ...

return {k: v for k, v in vars(Foo).items() if not k.startswith('_')}
```

Reply:
I changed to an approach where everything is manually added to a patches dictionary with a decorator.